File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.answer = None
    
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def postorderDFS(node: TreeNode) -> list[TreeNode]: # Post Order
            if node is None: # Node is empty
                return False
            left = postorderDFS(node.left) # Iterate left
            right = postorderDFS(node.right) # Iterate right
            mid = node == p or node == q # Check root

            if mid + left + right >= 2: # Check if subtree contains all nodes
                self.answer = node
            return mid or left or right # Return if at least p or q node exists
        
        postorderDFS(root)
        return self.answer

File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution 
{
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) 
    {
        // Tree does not exist
        if (root == null) return root;
        
        // Found an ancestor node (self)
        if (root == p || root == q) return root;
        
        // Recursively DFS the tree to find ancestors
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        
        // Determine which node to return as the common ancestor
        if (left != null && right != null) return root;
        if (left == null) return right;
        return left;
    }
}